staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go

```diff
@@ -0,0 +1,278 @@
+	// key here is dummy item in the queue to trigger file content processing.
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	var (
+		updatedEffectiveConfig bool
+		err                    error
+	)
+
+	// get context to close the new transformers.
+	ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
+
+	defer func() {
+		// TODO: increment success metric when updatedEffectiveConfig=true
+
+		if !updatedEffectiveConfig {
+			// avoid leaking if we're not using the newly constructed transformers (due to an error or them not being changed)
+			closeTransformers()
+		}
+		if err != nil {
+			// TODO: increment failure metric
+			utilruntime.HandleError(fmt.Errorf("error processing encryption config file %s: %v", d.filePath, err))
+			// add dummy item back to the queue to trigger file content processing.
+			d.queue.AddRateLimited(key)
+		}
+	}()
+
+	encryptionConfiguration, configChanged, err := d.processEncryptionConfig(ctx)
+	if err != nil {
+		return true
+	}
+	if !configChanged {
+		return true
+	}
+
+	// get healthz checks for all new KMS plugins.
+	if err = d.validateNewTransformersHealth(encryptionConfiguration.HealthChecks); err != nil {
+		return true
+	}
+
+	// update transformers.
+	// when reload=true there must always be one healthz check.
+	if len(encryptionConfiguration.HealthChecks) == 1 {
+		d.dynamicTransformers.Set(
+			encryptionConfiguration.Transformers,
+			closeTransformers,
+			encryptionConfiguration.HealthChecks[0],
+			encryptionConfiguration.KMSTimeout,
+		)
+
+		// update local copy of recent config content once update is successful.
+		d.lastLoadedEncryptionConfigHash = encryptionConfiguration.EncryptionFileContentHash
+		klog.V(2).InfoS("Loaded new kms encryption config content", "name", d.name)
+
+		updatedEffectiveConfig = true
+	} else {
+		err = fmt.Errorf("unexpected number of healthz checks: %d. Should have only one", len(encryptionConfiguration.HealthChecks))
+	}
+	return true
+}
+
+// loadEncryptionConfig processes the next set of content from the file.
+func (d *DynamicKMSEncryptionConfigContent) processEncryptionConfig(ctx context.Context) (
+	encryptionConfiguration *encryptionconfig.EncryptionConfiguration,
+	configChanged bool,
+	err error,
+) {
+	encryptionConfig, err := d.readEncryptionFileContent()
+	if err != nil {
+		return nil, false, err
+	}
+	if len(encryptionConfig) == 0 {
+		return nil, false, fmt.Errorf("encryption config file is empty %q", d.filePath)
+	}
+
+	// check if encryptionConfig is different from the current. Do nothing if they are the same.
+	if !d.hasEncryptionConfigChanged(encryptionconfig.ComputeEncryptionConfigHash(encryptionConfig)) {
+		return nil, false, nil
+	}
+
+	// this code path will only execute if reload=true. So passing true explicitly.
+	encryptionConfiguration, err = encryptionconfig.LoadEncryptionConfig(d.filePath, true, ctx.Done())
+
+	return encryptionConfiguration, true, err
```

Comment:
there's a time-of-check-time-of-use race here... we're reading the content and computing the hash above, but then LoadEncryptionConfig re-reads the file and recomputes the hash

if a file was written incrementally in place (not a good idea, but possible), our local read/hash could think the file changed, then the LoadEncryptionConfig could see the original content and hash

should we do this instead:
```go
if err != nil {
  return nil, false, err
}
return encryptionConfiguration, d.hasEncryptionConfigChanged(encryptionConfiguration.EncryptionFileContentHash), nil
```